xonsh/history/json.py

```diff
@@ -373,10 +381,19 @@ def flush(self, at_exit=False):
         """
         if len(self.buffer) == 0:
             return
+
+        def skip(num):
+            self._skipped += num
+
         hf = JsonHistoryFlusher(
-            self.filename, tuple(self.buffer), self._queue, self._cond, at_exit=at_exit
+            self.filename,
+            tuple(self.buffer),
+            self._queue,
+            self._cond,
+            at_exit=at_exit,
+            skip=skip,
         )
-        self.buffer.clear()
+        self.buffer = []
```

Comment:
Why make a new list here?

Reply:
See https://github.com/xonsh/xonsh/issues/3562#issuecomment-625258720.

Reply:
Passing a list as an argument does not deepcopy the list, so you can't just pass a list to a background thread for processing and clear it immediately. If you do, there's nothing left for processing on the background thread - just an empty list.